### ISSUE TITLE
Latest versions of ESPv2 do not work with manually specified service account key file

### ISSUE THREAD

**Issue by nareddyt** [GoogleCloudPlatform/esp-v2#238]
_Originally posted by @someone1 in https://github.com/GoogleCloudPlatform/esp-v2/issues/185#issuecomment-662796933_

@nareddyt - should this be working now with 2.14? It works when run on Cloud Run but not locally. It seems to get a lot further, actually downloading the config and listening to requests and the previous snake_case error message no longer appears. However, I still only get a single request before the entire container locks up, and it never reaches my backend:

HTTP Response (Status Code 500):
```
{"code":500,"message":"INTERNAL:Calling Google Service Control API failed with: 401 and body: \b\u0010\u0012Request had invalid authentication credentials. Expected OAuth 2 access token, login cookie or other valid authentication credential. See https://developers.google.com/identity/sign-in/web/devconsole-project."}
```

Relevant logs around the request:
```
D0723 02:03:33.267 53 envoy] [53][http][external/envoy/source/common/http/conn_manager_impl.cc:261] [C3] new stream
D0723 02:03:33.267 53 envoy] [53][http][external/envoy/source/common/http/conn_manager_impl.cc:808] [C3][S14536850401322214170] request headers complete (end_stream=true):\n\':authority\', \'localhost:8089\'\n\':path\', \'/v1alpha/accounts?key=<removed>\'\n\':method\', \'GET\'\n\'connection\', \'keep-alive\'\n\'pragma\', \'no-cache\'\n\'cache-control\', \'no-cache\'\n\'authorization\', \'Bearer <removed>\'

D0723 02:03:33.267 53 envoy] [53][http][external/envoy/source/common/http/conn_manager_impl.cc:1377] [C3][S14536850401322214170] request end stream
D0723 02:03:33.267 53 envoy] [53][jwt][external/envoy/source/extensions/filters/http/jwt_authn/filter.cc:127] Called Filter : setDecoderFilterCallbacks
D0723 02:03:33.267 53 envoy] [53][filter][src/envoy/http/path_matcher/filter.cc:81] matched operation: v1alpha.web_api_dot_project_id_appspot_com.Web_ListAccounts
D0723 02:03:33.267 53 envoy] [53][jwt][external/envoy/source/extensions/filters/http/jwt_authn/filter.cc:49] Called Filter : decodeHeaders
D0723 02:03:33.267 53 envoy] [53][jwt][bazel-out/k8-opt/bin/external/envoy/source/extensions/filters/http/jwt_authn/_virtual_includes/filter_config_interface/extensions/filters/http/jwt_authn/filter_config.h:117] use filter state value v1alpha.web_api_dot_project_id_appspot_com.Web_ListAccounts to find verifier.
D0723 02:03:33.267 53 envoy] [53][jwt][external/envoy/source/extensions/filters/http/jwt_authn/extractor.cc:188] extract authorizationBearer 
D0723 02:03:33.267 53 envoy] [53][jwt][external/envoy/source/extensions/filters/http/jwt_authn/extractor.cc:188] extract x-goog-iap-jwt-assertion
D0723 02:03:33.267 53 envoy] [53][jwt][external/envoy/source/extensions/filters/http/jwt_authn/authenticator.cc:127] auth0_jwk: JWT authentication starts (allow_failed=false), tokens size=1
D0723 02:03:33.267 53 envoy] [53][jwt][external/envoy/source/extensions/filters/http/jwt_authn/authenticator.cc:138] auth0_jwk: startVerify: tokens size 1
D0723 02:03:33.267 53 envoy] [53][jwt][external/envoy/source/extensions/filters/http/jwt_authn/authenticator.cc:149] auth0_jwk: Verifying JWT token of issuer https://<project-id>.auth0.com/
D0723 02:03:33.267 53 envoy] [53][filter][external/envoy/source/extensions/filters/http/common/jwks_fetcher.cc:55] fetch pubkey from [uri = https://<project-id>.auth0.com/.well-known/jwks.json]: start
D0723 02:03:33.267 53 envoy] [53][router][external/envoy/source/common/router/router.cc:426] [C0][S12502690874350086838] cluster \'<project-id>.auth0.com:443\' match for URL \'/.well-known/jwks.json\'
D0723 02:03:33.267 53 envoy] [53][router][external/envoy/source/common/router/router.cc:583] [C0][S12502690874350086838] router decoding headers:\n\':path\', \'/.well-known/jwks.json\'\n\':authority\', \'<project-id>.auth0.com\'\n\':method\', \'GET\'\n\':scheme\', \'https\'\n\'x-envoy-internal\', \'true\'\n\'x-forwarded-for\', \'172.18.0.8\'\n\'x-envoy-expected-rq-timeout-ms\', \'30000\'

D0723 02:03:33.267 53 envoy] [53][pool][external/envoy/source/common/http/conn_pool_base.cc:71] queueing request due to no available connections
D0723 02:03:33.267 53 envoy] [53][pool][external/envoy/source/common/conn_pool/conn_pool_base.cc:53] creating a new connection
D0723 02:03:33.268 53 envoy] [53][client][external/envoy/source/common/http/codec_client.cc:35] [C4] connecting
D0723 02:03:33.268 53 envoy] [53][connection][external/envoy/source/common/network/connection_impl.cc:753] [C4] connecting to 35.165.199.195:443
D0723 02:03:33.268 53 envoy] [53][connection][external/envoy/source/common/network/connection_impl.cc:769] [C4] connection in progress
D0723 02:03:33.268 53 envoy] [53][jwt][external/envoy/source/extensions/filters/http/jwt_authn/filter.cc:76] Called Filter : decodeHeaders Stop
D0723 02:03:33.333 53 envoy] [53][connection][external/envoy/source/common/network/connection_impl.cc:616] [C4] connected
D0723 02:03:33.333 53 envoy] [53][connection][external/envoy/source/extensions/transport_sockets/tls/ssl_socket.cc:191] [C4] handshake expecting read
D0723 02:03:33.399 53 envoy] [53][connection][external/envoy/source/extensions/transport_sockets/tls/ssl_socket.cc:191] [C4] handshake expecting read
D0723 02:03:33.399 53 envoy] [53][connection][external/envoy/source/extensions/transport_sockets/tls/ssl_socket.cc:191] [C4] handshake expecting read
D0723 02:03:33.399 53 envoy] [53][connection][external/envoy/source/extensions/transport_sockets/tls/ssl_socket.cc:191] [C4] handshake expecting read
D0723 02:03:33.401 53 envoy] [53][connection][external/envoy/source/extensions/transport_sockets/tls/ssl_socket.cc:191] [C4] handshake expecting read
D0723 02:03:33.401 53 envoy] [53][connection][external/envoy/source/extensions/transport_sockets/tls/ssl_socket.cc:191] [C4] handshake expecting read
D0723 02:03:33.466 53 envoy] [53][connection][external/envoy/source/extensions/transport_sockets/tls/ssl_socket.cc:176] [C4] handshake complete
D0723 02:03:33.466 53 envoy] [53][client][external/envoy/source/common/http/codec_client.cc:73] [C4] connected
D0723 02:03:33.466 53 envoy] [53][pool][external/envoy/source/common/conn_pool/conn_pool_base.cc:146] [C4] attaching to next request
D0723 02:03:33.466 53 envoy] [53][pool][external/envoy/source/common/conn_pool/conn_pool_base.cc:73] [C4] creating stream
D0723 02:03:33.466 53 envoy] [53][router][external/envoy/source/common/router/upstream_request.cc:342] [C0][S12502690874350086838] pool ready
D0723 02:03:33.513 53 envoy] [53][conn_handler][external/envoy/source/server/connection_handler_impl.cc:422] [C5] new connection
D0723 02:03:33.545 53 envoy] [53][router][external/envoy/source/common/router/router.cc:1153] [C0][S12502690874350086838] upstream headers complete: end_stream=false
D0723 02:03:33.545 53 envoy] [53][http][external/envoy/source/common/http/async_client_impl.cc:99] async http request response headers (end_stream=false):\n\':status\', \'200\'\n\'date\', \'Thu, 23 Jul 2020 02:03:33 GMT\'\n\'content-type\', \'application/json; charset=utf-8\'\n\'content-length\', \'3160\'\n\'connection\', \'keep-alive\'\n\'server\', \'nginx\'\n\'vary\', \'Accept-Encoding,Accept-Encoding\'\n\'ot-tracer-spanid\', \'7ce79021366dfbbf\'\n\'ot-tracer-traceid\', \'072b59c46a854402\'\n\'ot-tracer-sampled\', \'true\'\n\'ot-baggage-auth0-request-id\', \'8b8f630012aa4e2c4e692a80\'\n\'x-auth0-requestid\', \'85ce479a57d43747ed67\'\n\'access-control-allow-origin\', \'*\'\n\'access-control-allow-credentials\', \'false\'\n\'x-ratelimit-limit\', \'20\'\n\'x-ratelimit-remaining\', \'19\'\n\'x-ratelimit-reset\', \'1595469814\'\n\'cache-control\', \'public, max-age=15, stale-while-revalidate=15, stale-if-error=15\'\n\'etag\', \'W/\"c58-WlyAZ0fSEWSGQDEN8saHRcVzjpw\"\'\n\'x-cache-status\', \'EXPIRED\'\n\'x-envoy-upstream-service-time\', \'277\'

D0723 02:03:33.545 53 envoy] [53][client][external/envoy/source/common/http/codec_client.cc:105] [C4] response complete
D0723 02:03:33.545 53 envoy] [53][filter][external/envoy/source/extensions/filters/http/common/jwks_fetcher.cc:71] onSuccess: fetch pubkey [uri = https://<project-id>.auth0.com/.well-known/jwks.json]: success
D0723 02:03:33.545 53 envoy] [53][filter][external/envoy/source/extensions/filters/http/common/jwks_fetcher.cc:78] onSuccess: fetch pubkey [uri = https://<project-id>.auth0.com/.well-known/jwks.json]: succeeded
D0723 02:03:33.546 53 envoy] [53][jwt][external/envoy/source/extensions/filters/http/jwt_authn/authenticator.cc:267] auth0_jwk: JWT token verification completed with: OK
D0723 02:03:33.546 53 envoy] [53][jwt][external/envoy/source/extensions/filters/http/jwt_authn/filter.cc:87] Called Filter : check complete OK
D0723 02:03:33.546 53 envoy] [53][filter][src/envoy/http/service_control/filter.cc:46] Called ServiceControl Filter : decodeHeaders
D0723 02:03:33.546 53 envoy] [53][filter][src/envoy/http/service_control/service_control_call_impl.cc:182] Sending check : service_name: \"web-api-dot-<project-id>.appspot.com\"\noperation {\n  operation_id: \"a50ce11c-62fc-471c-8183-358ed0284212\"\n  operation_name: \"v1alpha.web_api_dot_project_id_appspot_com.Web_ListAccounts\"\n  consumer_id: \"api_key:<removed>\"\n  start_time {\n    seconds: 1595469813\n    nanos: 546096908\n  }\n  end_time {\n    seconds: 1595469813\n    nanos: 546096908\n  }\n  labels {\n    key: \"servicecontrol.googleapis.com/caller_ip\"\n    value: \"172.18.0.1\"\n  }\n  labels {\n    key: \"servicecontrol.googleapis.com/referer\"\n    value: \"http://localhost:8080/accounts/manage\"\n  }\n  labels {\n    key: \"servicecontrol.googleapis.com/service_agent\"\n    value: \"ESPv2/2.14.0\"\n  }\n  labels {\n    key: \"servicecontrol.googleapis.com/user_agent\"\n    value: \"ESPv2\"\n  }\n}\nservice_config_id: \"2020-07-16r1\"

D0723 02:03:33.546 53 envoy] [53][filter][src/envoy/http/service_control/http_call.cc:322] Service Control remote call: Check is created
D0723 02:03:33.546 53 envoy] [53][filter][src/envoy/http/service_control/http_call.cc:201] http call from [uri = https://servicecontrol.googleapis.com/v1/services/web-api-dot-<project-id>.appspot.com:check]: start
D0723 02:03:33.546 53 envoy] [53][router][external/envoy/source/common/router/router.cc:426] [C0][S312826532313055775] cluster \'service-control-cluster\' match for URL \'/v1/services/web-api-dot-<project-id>.appspot.com:check\'
D0723 02:03:33.546 53 envoy] [53][router][external/envoy/source/common/router/router.cc:583] [C0][S312826532313055775] router decoding headers:\n\':path\', \'/v1/services/web-api-dot-<project-id>.appspot.com:check\'\n\':authority\', \'servicecontrol.googleapis.com\'\n\':method\', \'POST\'\n\':scheme\', \'https\'\n\'content-length\', \'505\'\n\'authorization\', \'Bearer <removed>\'

D0723 02:03:33.546 53 envoy] [53][pool][external/envoy/source/common/http/conn_pool_base.cc:71] queueing request due to no available connections
D0723 02:03:33.546 53 envoy] [53][pool][external/envoy/source/common/conn_pool/conn_pool_base.cc:53] creating a new connection
D0723 02:03:33.546 53 envoy] [53][client][external/envoy/source/common/http/codec_client.cc:35] [C6] connecting
D0723 02:03:33.546 53 envoy] [53][connection][external/envoy/source/common/network/connection_impl.cc:753] [C6] connecting to 172.217.10.234:443
D0723 02:03:33.546 53 envoy] [53][connection][external/envoy/source/common/network/connection_impl.cc:769] [C6] connection in progress
D0723 02:03:33.546 53 envoy] [53][filter][src/envoy/http/service_control/filter.cc:64] Called ServiceControl filter : Stop
D0723 02:03:33.546 53 envoy] [53][pool][external/envoy/source/common/http/http1/conn_pool.cc:49] [C4] response complete
D0723 02:03:33.546 53 envoy] [53][pool][external/envoy/source/common/conn_pool/conn_pool_base.cc:98] [C4] destroying stream: 0 remaining
D0723 02:03:33.564 53 envoy] [53][connection][external/envoy/source/common/network/connection_impl.cc:616] [C6] connected
D0723 02:03:33.564 53 envoy] [53][connection][external/envoy/source/extensions/transport_sockets/tls/ssl_socket.cc:191] [C6] handshake expecting read
D0723 02:03:33.589 53 envoy] [53][connection][external/envoy/source/extensions/transport_sockets/tls/ssl_socket.cc:191] [C6] handshake expecting read
D0723 02:03:33.589 53 envoy] [53][connection][external/envoy/source/extensions/transport_sockets/tls/ssl_socket.cc:191] [C6] handshake expecting read
D0723 02:03:33.589 53 envoy] [53][connection][external/envoy/source/extensions/transport_sockets/tls/ssl_socket.cc:191] [C6] handshake expecting read
D0723 02:03:33.600 53 envoy] [53][connection][external/envoy/source/extensions/transport_sockets/tls/ssl_socket.cc:176] [C6] handshake complete
D0723 02:03:33.600 53 envoy] [53][client][external/envoy/source/common/http/codec_client.cc:73] [C6] connected
D0723 02:03:33.600 53 envoy] [53][pool][external/envoy/source/common/conn_pool/conn_pool_base.cc:146] [C6] attaching to next request
D0723 02:03:33.600 53 envoy] [53][pool][external/envoy/source/common/conn_pool/conn_pool_base.cc:73] [C6] creating stream
D0723 02:03:33.600 53 envoy] [53][router][external/envoy/source/common/router/upstream_request.cc:342] [C0][S312826532313055775] pool ready
D0723 02:03:33.629 53 envoy] [53][router][external/envoy/source/common/router/router.cc:1153] [C0][S312826532313055775] upstream headers complete: end_stream=false
D0723 02:03:33.629 53 envoy] [53][http][external/envoy/source/common/http/async_client_impl.cc:99] async http request response headers (end_stream=false):\n\':status\', \'401\'\n\'www-authenticate\', \'Bearer <removed>\'

D0723 02:03:33.629 53 envoy] [53][client][external/envoy/source/common/http/codec_client.cc:105] [C6] response complete
D0723 02:03:33.630 53 envoy] [53][filter][src/envoy/http/service_control/http_call.cc:101] http call response status code: 401, body: \010\020\022\320\001Request had invalid authentication credentials. Expected OAuth 2 access token, login cookie or other valid authentication credential. See https://developers.google.com/identity/sign-in/web/devconsole-project.
E0723 02:03:33.630 53 envoy] [53][filter][src/envoy/http/service_control/client_cache.cc:144] Failed to call check, error: UNAUTHENTICATED:Calling Google Service Control API failed with: 401 and body: \010\020\022\320\001Request had invalid authentication credentials. Expected OAuth 2 access token, login cookie or other valid authentication credential. See https://developers.google.com/identity/sign-in/web/devconsole-project., str body: \010\020\022\320\001Request had invalid authentication credentials. Expected OAuth 2 access token, login cookie or other valid authentication credential. See https://developers.google.com/identity/sign-in/web/devconsole-project.
E0723 02:03:43.942451       6 rollout_id_change_detector.go:65] error occurred when checking new rollout id, fail to fetch new rollout id, http call to POST https://servicecontrol.googleapis.com/v1/services/web-api-dot-<project-id>.appspot.com:report returns not 200 OK: 403 Forbidden
```

I can open a new issue if this is unrelated - and I'd be happy to share my entire log privately if it helps any. I'm using the same ESPv2 startup options as mentioned earlier in this issue. It continues to work fine with ESPv2.7.0.

I should mention, the 401/403 error messages may be symptomatic of an error earlier in the stack. The following works fine with the same service account file so I'm pretty sure it has permissions:

```sh
curl -s \
    -H "Authorization: Bearer $(gcloud auth application-default print-access-token)" \
    -H "Content-Type: application/json" \
    https://servicecontrol.googleapis.com/v1/services/web-api-dot-<project-id>.appspot.com:check \
    --data '{
        "serviceConfigId": "2020-07-16r1", 
        operation: {
            "operationId": "<removed>",
            "consumerId":  "api_key:<removed>",
            "operationName": "<removed>",
            "startTime": "2020-07-22T22:01:23.045123456Z",
            "endTime": "2020-07-22T22:01:23.045123456Z",
        }
    }'
```

**Comment by nareddyt** [comment]
Hi @someone1, I'm fairly certain this is a permission issue.

Previous versions of ESP sometimes allow requests through even if the service account is missing permissions to call Service Control. In ESPv2 2.8.0 and 2.9.0, we resolved most of these permissions problems and display clear error messages to the client. That was part of PRs #110 and #122. You are seeing the error now after updating to 2.14.0.

In fact, take a look at your Cloud Endpoint service that is running ESP 2.7.0. If you look through the Cloud Console, I doubt you'll see and metrics or logs, even though the request is allowed through.

To resolve this: Take a look at bullet number 4 in the [Building a new ESPv2 Beta Image](https://cloud.google.com/endpoints/docs/openapi/get-started-cloud-run#configure_esp) section of the Cloud Run tutorial. You need to ensure the service account has `roles/servicemanagement.serviceController`.

**Comment by someone1** [comment]
Thanks for the reply! I do have that role assigned to the service account, and as aforementioned, the manual curl request works without issue. Here are the roles assigned (I even granted owner on the project in an act of desperation after [reading the docs](https://cloud.google.com/service-infrastructure/docs/service-control/access-control)):

![image](https://user-images.githubusercontent.com/1321933/88323191-dc178380-ccef-11ea-8a6b-584fe26d3ef8.png)

**Comment by someone1** [comment]
Taking a look at the JWT from the log: 

```
D0723 02:01:44.484 30 envoy] [30][filter][src/envoy/token/sa_token_generator.cc:56] Generated JWT from service account: <REMOVED>
```
jwt.io reports:

```
HEADER:ALGORITHM & TOKEN TYPE

{
  "alg": "RS256",
  "typ": "JWT",
  "kid": "d489deab6e41352706398d1c086caf445224d411"
}
PAYLOAD:DATA

{
  "iss": "<service_account>",
  "sub": "<service_account>",
  "aud": "https://servicecontrol.googleapis.com/v1/services//google.api.servicecontrol.v1.ServiceController",
  "iat": 1595469704,
  "exp": 1595473304
}
```

However, the audience seems a little off? Retrying the curl with the JWT from the logs, it does fail:

```
{
  "error": {
    "code": 401,
    "message": "Request had invalid authentication credentials. Expected OAuth 2 access token, login cookie or other valid authentication credential. See https://developers.google.com/identity/sign-in/web/devconsole-project.",
    "status": "UNAUTHENTICATED"
  }
}
```

but using the access token from `gcloud auth application-default print-access-token` works fine. I validated that the `issued_to` of the token matches the `client_id` of the service account json with:

```
curl -X POST https://www.googleapis.com/oauth2/v1/tokeninfo?access_token=$(gcloud auth application-default print-access-token)
```

**Comment by TAOXUY** [comment]
Hi Prateek,

The double `/`  in the audience looks like terrible.  
""aud": "https://servicecontrol.googleapis.com/v1/services//google.api.servicecontrol.v1.ServiceController","

Let me test it out and gets back to you.

**Comment by nareddyt** [comment]
Ah, I misread your original post. It works when running on Cloud Run but not locally.

Can you briefly describe how you're running ESPv2 locally? Which doc did you follow, and how are you providing the service account key file to ESP?

To be clear, if you run locally, you need to provide [a service account key manually](https://github.com/GoogleCloudPlatform/esp-v2/blob/db1ec6b904c9e0a72e3d05708c0eb62a83640f1b/docker/generic/start_proxy.py#L493), because you don't have Google Metadata server on our local computer. So if you can clarify how you run it locally, that will help.

JWT issue is weird, I'll let @TAOXUY look into it.

**Comment by someone1** [comment]
Sure - I use VS Code w/ Remote Containers (on a Windows host) and have a `docker-compose.yaml` with the following:

```yaml
  esp:
    image: gcr.io/endpoints-release/endpoints-runtime:2.7 # TODO: Update when bugs are fixed
    restart: unless-stopped
    container_name: esp
    ports:
      - "127.0.0.1:8089:8089"
    volumes:
      - ${USERPROFILE}/.config/.google-application-credentials.json:/.google-application-credentials.json:ro
    environment:
      - GOOGLE_APPLICATION_CREDENTIALS=/.google-application-credentials.json
    command:
      - "--service=<service name>"
      - "--rollout_strategy=managed"
      - "--listener_port=8089"
      - "--backend=http://dev:8081"
      - "--service_account_key=/.google-application-credentials.json"
      - "--cors_preset=basic"
      - "--cors_allow_origin=http://localhost:8080"
      - "--non_gcp"
      - "--enable_debug"
    links:
      - dev
```

I setup a `local-dev` service account I share with my team to develop locally with which is the `${USERPROFILE}/.config/.google-application-credentials.json` file mentioned above. I set the permissions as appropriate for the various services we may need to access locally to a shared developer test project.

My API backend runs on port 8081 while my frontend app runs on port 8080 and is configured to send API requests to localhost:8089 when run locally.

I didn't really follow any docs - I mostly took the setup I had from ESPv1 and read through the proxy python script and config manager Go code to see which options are supported on ESPv2 (or reasoned about which would be useful for local dev)

**Comment by nareddyt** [comment]
Jason, does #243 reproduce this?

**Comment by TAOXUY** [comment]
Yes, I reproduced this error. 

The first bug is the trailing slash but after fixing that, the access token generated is still problematic. Maybe we need tweak our lib a little bit. Anyway, I directly go with implementing the local token server on configmanager side.

**Comment by nareddyt** [comment]
Ah I see the trailing slash error.

Confused why the generated token is still problematic, as it looks like it's working on version 2.7

Anyways, that's fine if implementing it on the config manager side is easier.

**Comment by nareddyt** [comment]
I believe #251 should fix this. It also adds an e2e test using a service account file.

We'll make a new release 2.15.0 this week.

**Comment by someone1** [comment]
The [2.15.0 release](https://github.com/GoogleCloudPlatform/esp-v2/releases/tag/v2.15.0) mentions it was produced on the commit right before the #251 merge (38f0610), but the [commit list](https://github.com/GoogleCloudPlatform/esp-v2/commits/v2.15.0) for the release includes it - did it actually make it or should I wait until 2.16.0?

**Comment by TAOXUY** [comment]
> The [2.15.0 release](https://github.com/GoogleCloudPlatform/esp-v2/releases/tag/v2.15.0) mentions it was produced on the commit right before the #251 merge ([38f0610](https://github.com/GoogleCloudPlatform/esp-v2/commit/38f0610d80ecc99cf3833f20f860b9d08c9fa366)), but the [commit list](https://github.com/GoogleCloudPlatform/esp-v2/commits/v2.15.0) for the release includes it - did it actually make it or should I wait until 2.16.0?

Yes, this change is not included in the released image and we will release it in 2.16.0. For now, if you are urgent, you can try out gcr.io/cloudesf-testing/apiproxy-serverless:1c363295851b106bd57305d71e9fec3abe7457a.
